### ISSUE TITLE
SVG's rendering is prone to XSS

### ISSUE THREAD

**Issue by maudov** [mikitex70/redmine_drawio#102]
Hi,

SVG's rendering is prone to XSS in redmine_drawio.
steps to reproduce : 
1. upload an SVG containing a script to a wiki page and attach it ( {{drawio_attach(...)}})
```
<svg height="100" width="100">

  <script>alert('xss')</script>

  <circle cx="50" cy="50" r="40" stroke="black" stroke-width="3" fill="red" />

  Sorry, your browser does not support inline SVG.  

</svg> 
```

2. go to the wiki page

![image](https://user-images.githubusercontent.com/90309140/134655374-be2b7093-dd9f-4353-b542-aa595fe8677f.png)

there are several approaches to fix this issue:
- set the CSP policy to prevent inline javascript execution
- insert the SVG in iframe tag with the sandbox attribute
- insert the SVG as base64 in an img tag

The latter one is probalby the easiest fix for redmine_drawio, here is an example:
https://github.com/mathieu-mbru/redmine_drawio/commit/50ac2d9753187ceb8fe487dd9d6da7bfd18a8fc4

More information about XSS here:
https://owasp.org/www-community/attacks/xss/

An interesting article about SVG XSS defence scenarios:
https://digi.ninja/blog/svg_xss.php

**Comment by mikitex70** [comment]
Thanks @maudov, I will fix it as soon a possible!

**Comment by mikitex70** [comment]
Hi @maudov, the security problem should now be fixed.
I've opted for removing javascripts from the svg diagram, because working with base64 was causing some issue when the diagram was re-edited before reloading the page.
Thanks again for reporting it.

**Comment by maudov** [comment]
Hi @mikitex70,
thank you for your fast reaction, unfortunately there are many ways to bypass the fix (i.e : <sCriPt>alert('coucou')<sCript>).

A blacklisting approach is very difficult.
[here](https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html) you can have a glimpse of all the XSS attacks used to bypass XSS filtering.

I will try to fix the re-editing issue you had the base64 approach.

**Comment by mikitex70** [comment]
Ok, I will try to use the base64.
From your experience the SVG is still navigable? We use diagrams with links to other pages in Redmine and if the use of the base64 encoding breaks this navigation it maybe  problem.

**Comment by mikitex70** [comment]
Hi @maudov, I've pushed in the `develop` branch a fix for the XSS svg issues. Can yo give it a try?
Basically I've added a new plugin configuration option, `Enable SVG diagrams`, which is by default unchecked.

If unchecked, svg diagrams are rendered as inline base64 svg+xml images, as your suggestion. They are fully editable (thanks for your PR). In this case hyperlinks don't work.

If checked, svg diagrams are rendered as before (svg inserted directly in the page), and the hyperlinks are working. The XSS issues are mitigated by filtering out (some) of the potentially dangerous scripts.

So, the Redmine administrator is responsible of choosing the level of security: for intranet maybe the svg can be safely enabled, for internet it's better to unmark the check and let render them as as base64 images.

I hope this solution can cover all needs, keeping also a compatibility with previous releases where svg navigation is important.

As alternative, one can use diagrams in xml format, which are rendered as svg but they don't allow scripting.

**Comment by maudov** [comment]
Hi @mikitex70,
It seems to be a good solution to me. I will give it a try as soon as I can.

**Comment by maudov** [comment]
hi @mikitex70, I just tried the develop branch and I think there is a bug when editing an SVG (when SVG diagrams are enabled).
After saving the new SVG is not refreshed and I need to refresh my browser.
Except this minor bug everything seems to work fine and the solution seems to cover all needs.
Thank you for the fix!

**Comment by mikitex70** [comment]
Thanks for your tests, this this bug had escaped me.
It is now fixed, and I've released version `1.2.0`.
Thank you very much for reporting the security issue.
